source/Detector/Amptek/AmptekSDD123Detector.cpp

```diff
@@ -0,0 +1,656 @@
+	:QObject(parent)
+{
+	spectrumReceiver_ = 0;
+	name_ = name;
+	basePVName_ = basePVName;
+	dataType_ = dataType;
+	bufferSize_ = bufferSize;
+
+	configurationParametersInitialized_ = false;
+
+	fastCounts_ = -1;
+	slowCounts_ = -1;
+	generalPurposeCounter_ = -1;
+	accumulationTime_ = -1;
+	liveTime_ = -1;
+	realTime_ = -1;
+	firmwareMajorVersion_ = -1;
+	firmwareMinorVersion_ = -1;
+	fpgaMajorVersion_ = -1;
+	fpgaMinorVersion_ = -1;
+	serialNumber_ = -1;
+	highVoltage_ = 272727272727;
+	detectorTemperature_ = -1;
+	boardTemperature_ = 2727272727;
+	presetRealTimeReached_ = false;
+	autoFastThresholdLocked_ = false;
+	mcaEnabled_ = false;
+	presetCountReached_ = false;
+	oscilloscopeDataReady_ = false;
+	unitConfigured_ = false;
+	autoInputOffsetSearching_ = false;
+	mcsFinished_ = false;
+	using20MHz_ = false;
+	using80MHz_ = false;
+	clockAutoSet_ = false;
+	pc5DetectedAtPowerUp_ = false;
+	pc5HVNegative_ = false;
+	pc5HVPositive_ = false;
+	pc5Preamp5V_ = false;
+	pc5Preamp8p5V_ = false;
+
+	presetRealTimeReachedInitialized_ = false;
+	autoFastThresholdLockedInitialized_ = false;
+	mcaEnabledInitialized_ = false;
+	presetCountReachedInitialized_ = false;
+	oscilloscopeDataReadyInitialized_ = false;
+	unitConfiguredInitialized_ = false;
+	autoInputOffsetSearchingInitialized_ = false;
+	mcsFinishedInitialized_ = false;
+	using20MHzInitialized_ = false;
+	using80MHzInitialized_ = false;
+	clockAutoSetInitialized_ = false;
+	pc5DetectedAtPowerUpInitialized_ = false;
+	pc5HVNegativeInitialized_ = false;
+	pc5HVPositiveInitialized_ = false;
+	pc5Preamp5VInitialized_ = false;
+	pc5Preamp8p5VInitialized_ = false;
+
+	deviceID_ = -1;
+
+	connect(this, SIGNAL(postSpectrumPacketEventReceived(QStringList)), this, SLOT(onPostSpectrumPacketEventReceived(QStringList)));
+}
+
+bool AmptekSDD123Detector::event(QEvent *e){
+	if(e->type() == (QEvent::Type)AmptekEventDefinitions::SpectrumPacketEvent){
+		onSpectrumPacketEventReceived((AmptekSpectrumPacketEvent*)e);
+		e->accept();
+		return true;
+	}
+	else if(e->type() == (QEvent::Type)AmptekEventDefinitions::ConfigurationReadbackEvent){
+		onConfigurationReadbackEventReceived((AmptekConfigurationReadbackEvent*)e);
+	}
+
+	return QObject::event(e);
+}
+
+void AmptekSDD123Detector::onSpectrumPacketEventReceived(AmptekSpectrumPacketEvent* event)
+{
+
+	QByteArray spectrumByteArray = event->spectrumByteArray_;
+	int channelCount = event->channelCount_;
+	QByteArray statusDataArray = event->statusDataArray_;
+
+	AMDSFlatArray *spectrumArray = readSpectrumData(spectrumByteArray, channelCount);
+	readStatusData(statusDataArray);
+
+	if(spectrumReceiver_){
+
+		AMDSDwellStatusData statusData(fastCounts(), slowCounts(), detectorTemperature(), accumulationTime(), liveTime(), realTime(), generalPurposeCounter(), event->dwellStartTime_, event->dwellEndTime_, event->dwellReplyTime_);
+
+		AmptekSpectrumEvent *spectrumEvent = new AmptekSpectrumEvent();
```

Comment:
Since these are events, I'm assuming that the event handler handles the memory management of these once they're done?